fv3core/stencils/map_single.py

```diff
@@ -93,257 +90,175 @@ def lagrangian_contributions(
+                domain=domain,
+            )
 
+            self._set_remapped_quantity(
+                q1,
+                self._q2_adds,
+                origin=(origin[0], origin[1], k_eul),
+                domain=(domain[0], domain[1], 1),
+            )
 
-def lagrangian_contributions_stencil(
-    q1: FloatField,
-    pe1: FloatField,
-    pe2: FloatField,
-    q4_1: FloatField,
-    q4_2: FloatField,
-    q4_3: FloatField,
-    q4_4: FloatField,
-    dp1: FloatField,
-    i1: int,
-    i2: int,
-    j1: int,
-    j2: int,
-    kord: int,
-    origin: Tuple[int, int, int],
-    domain: Tuple[int, int, int],
-):
-    # A stencil with a loop over k2:
-    km = spec.grid.npz
-    shape2d = pe2.shape[0:2]
-    q2_adds = utils.make_storage_from_shape(shape2d, cache_key="map_single_q2_adds")
-    ptop = utils.make_storage_from_shape(shape2d, cache_key="map_single_ptop")
-    pbot = utils.make_storage_from_shape(shape2d, cache_key="map_single_pbot")
 
-    for k_eul in range(km):
+class MapSingle:
+    """
+    Fortran name is map_single, test class is Map1_PPM_2d
+    """
 
-        # TODO (olivere): This is hacky
-        # merge with subsequent stencil when possible
-        set_eulerian_pressures(
-            pe2,
-            ptop,
-            pbot,
-            origin=(origin[0], origin[1], k_eul),
-            domain=(domain[0], domain[1], 1),
-        )
+    def __init__(self, namelist: SimpleNamespace):
+        self._grid = spec.grid
+        shape = self._grid.domain_shape_full(add=(1, 1, 1))
+        origin = self._grid.compute_origin()
+
+        self.dp1 = utils.make_storage_from_shape(shape, origin=origin)
+        self.q4_1 = utils.make_storage_from_shape(shape, origin=origin)
+        self.q4_2 = utils.make_storage_from_shape(shape, origin=origin)
+        self.q4_3 = utils.make_storage_from_shape(shape, origin=origin)
+        self.q4_4 = utils.make_storage_from_shape(shape, origin=origin)
+
+        self.lagrangian_contributions = LagrangianContributions()
+
+        self._used_kords: Tuple[int] = (namelist.kord_tm, namelist.kord_tr)
+        self._used_modes: Tuple[int] = (-2, -1, 1)
+
+        kord_mode_combos: List[Tuple[int]] = []
+        for kord in self._used_kords:
+            for mode in self._used_modes:
+                kord_mode_combos.append((kord, mode))
 
-        lagrangian_contributions(
+        self._remap_profiles = {pair: RemapProfile(*pair) for pair in kord_mode_combos}
+        self._set_dp = StencilWrapper(set_dp)
+
+    def __call__(
+        self,
+        q1: FloatField,
+        pe1: FloatField,
+        pe2: FloatField,
+        qs: FloatField,
+        mode: int,
+        i1: int,
+        i2: int,
+        j1: int,
+        j2: int,
+        kord: int,
+        qmin: float = 0.0,
+    ):
+        """
+        Compute x-flux using the PPM method.
+
+        Args:
+            q1 (in): Transported scalar
+            pe1 (in): ???
+            pe2 (out): ???
+            qs (out): ???
```

Comment:
qs is the field to be remapped on the deformed grid, while q1 is the remapped field on the Eulerian grid.

It probably makes sense to rename these variables for clarity, actually

Reply:
Awesome, thank you.